src/dataflow/src/sink/kafka.rs

```diff
@@ -416,10 +525,20 @@ where
         stream.scope().index().to_string(),
         Arc::clone(&shutdown_flag),
         activator,
+        write_frontier,
     );
 
+    // Keep track of whether this operator/worker ever received updates. We
+    // use this to control who should send continuous END progress records
+    // to Kafka below.
+    let mut received_data = false;
+
     let mut vector = Vec::new();
 
+    // keep the latest progress updates, if any, in order to update
+    // our internal state after the send loop
+    let mut progress_updates = Vec::new();
```

Comment:
I feel like this can be an `Option<Timestamp>` or something. ideally you would just call `maybe_update_progress` in the main loop but it seems like borrow checker doesn't like that

Reply:
Yes, this is for getting around borrow checking! :sweat_smile: 
I had an `Option` here first, but I think we can emit multiple `END` records in one invocation. And a `Vec` should be fine since we're reusing the allocation. Could still change it, of course.

Reply:
yeah but you only ever update the internal field to the max anyway right? I guess maybe another question to ask: is it ever legit for `maybe_update_progress` to not update the last_progress_ts? as far as I can understand that would only happen if we emit times out of order which we don't want to do?

Reply:
I had a future in mind where we do allow emitting timestamps out of order. But I can turn it into an `Option` for now. 👌